lib/_http_outgoing.js

```diff
@@ -431,7 +431,8 @@ function _storeHeader(firstLine, headers) {
   }
 
   if (!state.contLen && !state.te) {
-    if (!this._hasBody) {
+    if (!this._hasBody && (this.statusCode === 204 ||
+                           this.statusCode === 304)) {
```

Comment:
Hmm I think that if the intent is for this is fix the behavior with HEAD requests, we likely should check the method explicitly.

Reply:
I guess the short answer is that the `OutgoingMessage` type doesn't have a field for the method of the request that it's responding to and I was afraid adding that field would be a larger re-factor.  As far as I can tell, it only requires updating one location (specifically avoiding collision with `method` in `_http_client.js`.)

That said, the status code check mirrors earlier logic: https://github.com/nodejs/node/pull/34231/files#diff-286202fdbdd74ede6f5f5334b6176b5cR410

I did build and test a patch with the suggested change (https://github.com/josephhackman/node/commit/9880a03e14eed3465a082248ef1a920666ed3437), and if you feel that's better I'd be happy to cherry-pick that into this PR.

Reply:
@jasnell What do you think? Should I pull in josephhackman@9880a03 ?

Reply:
Yeah I think that's best